src/librustc_mir/build/matches/mod.rs

```diff
@@ -770,21 +783,139 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
+        // match place { Foo { .. } if { let tmp1 = & &mut place; inspect(*tmp1) }
+        //     => { let tmp2 = &mut place; feed(tmp2) }, ... }
+        // ```
+        //
+        // In short, any pattern binding will always look like *some*
+        // kind of `&T` within the guard at least in terms of how the
+        // MIR-borrowck views it, and this will ensure that guard
+        // expressions cannot mutate their the match inputs via such
+        // bindings. (It also ensures that guard expressions can at
+        // most *copy* values from such bindings; non-Copy things
+        // cannot be moved via pattern bindings in guard expressions.)
+        //
+        // ----
+        //
+        // Implementation notes (under assumption `autoref` is true).
+        //
+        // To encode the distinction above, we must inject the
+        // temporaries `tmp1` and `tmp2`.
+        //
+        // There are two cases of interest: binding by-value, and binding by-ref.
+        //
+        // 1. Binding by-value: Things are simple.
+        //
+        //    * Establishing `tmp1` creates a reference into the
+        //      matched place. This code is emitted by
+        //      bind_matched_candidate_for_guard.
+        //
+        //    * `tmp2` is only initialized "lazily", after we have
+        //      checked the guard. Thus, the code that can trigger
+        //      moves out of the candidate can only fire after the
+        //      guard evaluated to true. This initialization code is
+        //      emitted by bind_matched_candidate_for_arm.
+        //
+        // 2. Binding by-reference: Things are tricky.
+        //
+        //    * Here, the guard expression wants a `&&` or `&&mut`
+        //      into the original input. This means we need to borrow
+        //      a reference that we do not immediately have at hand
+        //      (because all we have is the places associated with the
+        //      match input itself; it is up to us to create a place
+        //      holding a `&` or `&mut` that we can then borrow).
+        //
+        //    * Therefore, when the binding is by-reference, we
+        //      *eagerly* introduce the binding for the arm body
+        //      (`tmp2`) and then borrow it (`tmp1`).
+        //
+        //    * This is documented with "NOTE tricky business" below.
+        //
+        // FIXME The distinction in how `tmp2` is initialized is
+        // currently encoded in a pretty awkward fashion; namely, by
+        // passing a boolean to bind_matched_candidate_for_arm_body
+        // indicating whether all of the by-ref bindings were already
+        // initialized.
+        //
+        // * Also: pnkfelix thinks "laziness" is natural; but since
+        //   MIR-borrowck did not complain with earlier (universally
+        //   eager) MIR codegen, laziness might not be *necessary*.
+
+        let autoref = self.hir.tcx().nll();
         if let Some(guard) = candidate.guard {
```

Comment:
Sorry for being late with this, but I feel that the "guard bit" mapping - figuring out whether a local should access the "arm" version or the dereference of the "guard" version - belongs more in HAIR than in MIR.

Reply:
Hmm. I don't have a strong opinion, but that seems reasonable. How do you envision it working? I could imagine having e.g. distinct HIR nodes for "reference variable from guard" or something...?

Reply:
What I was thinking of was having variables in the HAIR refer to an enum
```Rust
enum HairVar {
    Binding(ast::NodeId),
    GuardRef(ast::NodeId)
}
```

MIR construction does not quite care about the exact `NodeId` of variables, only the scope of their definitions etc., so this should work out.

Reply:
> distinct HIR nodes for "reference variable from guard"

The above option would mean that we put it right in HAIR construction. Of course, we could also put it in resolution (using 2 separate `NodeId`s), but I understand the implications of that less.

Reply:
(In case it wasn't clear, I meant "distinct HAIR nodes", I edited my comment -- but it sounds like we were thinking of similar things)

Reply:
(at this point I would prefer to land this PR in the short term and come back later to the question of whether the guard bit mapping should part of the HAIR rather than the MIR... I will open an issue for that task once this has landed though...)